### ISSUE TITLE
Documentation and example of webhook signature check (through rawBody)

### ISSUE THREAD

**Issue by oom-** [fastify/fastify#5491]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the feature has not already been requested

### 🚀 Feature Proposal

I'm currently aiming to verify the signature of a webhook route using the raw body data.

There is already a lot of issues around the subject since 2018: 
- [Support rawBody or override JSON parser](https://github.com/fastify/fastify/issues/707)
- [Stripe webhooks rawBody](https://github.com/fastify/fastify/issues/1965)
- [preHandler types are broken for version](https://github.com/fastify/fastify/issues/5125)

There is still no clear documentation regarding signature verification or how to obtain a raw body inside the route handler to perform webhook signature check.

I think that the best option would still just to have a `rawBody: true` option on the route as @mcollina suggested here: 
[issuecomment-619153284](https://github.com/fastify/fastify/issues/707#issuecomment-619153284)

This way, the raw body would appear only on selected routes to avoid unnecessary memory consumption.

### Motivation

Enhance Fastify's usability and create a more welcoming environment.

### Example

Payload: `{"msg": "test"}`;

```ts
fastify.post("/example", { rawBody: true }, (request, reply) => {
   console.log(request.raw.body); // "{ \"msg\": \"text\" }"
   console.log(request.body); // { msg: 'text' }
});
```

**Comment by metcoder95** [comment]
I believe it can enable a good set of use cases the fact that we allow to the `parser` being skipped on certain routes based on a flag.

+1 on this.

Although here I'd like to see what are we interested the most, the intention of skipping the parser or the intention of just have the raw-body. I'd personally see more useful the former as it enables more use cases (e.g. pipelining between body reading and signature verification).

**Comment by xr0master** [comment]
I noticed that guys are working on version 5 and planning changes to data parsing. This might be a good opportunity to add a body to the FastifyRequest object's raw property. In my opinion, it has to be under `raw`. I also believe that this will not require a lot of code.

I am aware of the `fastify-raw-body` npm package, but I feel it does not align with the current Fastify philosophy. In the end, this package was written many years ago.

**Comment by xr0master** [comment]
@mcollina definitely, but that's why it should be disabled by default, then there shouldn't be any difference in performance.

On the other hand, adding a `preParsing` hook for a specific router only requires a few lines of code, and maybe that's more than enough — something like the code below.

```js
preParsing: (request, reply, payload, done) => {
  let data = '';

  payload.on('data', (chunk: string) => {
    data += chunk;
  });

  payload.on('end', () => {
    request.raw.body = data;
  });

  done(null, payload);
},
```

**Comment by metcoder95** [comment]
It looks like a plugin, and the `fastify-raw-body` plugin already does that; why should be added to core if it can be a plugin?

**Comment by xr0master** [comment]
Plugins typically add new features rather than restoring initial behavior. It often comes down to deciding what belongs in the core versus what should be handled by plugins. However, I agree with you — it makes sense for this to be a plugin. It would be ideal if the plugin were maintained by the Fastify team under the @fastify scope.

The `fastify-raw-body` brings too many extra packages for a straightforward thing.

**Comment by gurgunday** [comment]
```js
  fastify.addContentTypeParser(
    "application/json",
    { parseAs: "buffer" },
    (request, payload, done) => {
      done(null, payload);
    },
  );
```

For things like Stripe, adding this parser gives you the raw `request.body` for your desired contentType

**Comment by oom-** [comment]
Yes the best would be to have it optionally enabled per route and as native option.
No plugins, no performance lost on the rest of the app.

**Comment by Eomm** [comment]
> The fastify-raw-body brings too many extra packages for a straightforward thing, IMO.

Don't bring misinformation here that people do not verify these sentences.

| fastify-raw-body deps | Note |
|---|---|
secure-json-parse | already in fastify: https://github.com/fastify/fastify/blob/54b1022839a85f32899932e237e1a5463bfd96da/package.json#L204
| fastify-plugin | Every fastify plugin has it
| raw-body | 24k https://www.npmjs.com/package/raw-body to be HTTP compliant

**Comment by mahmoodhamdi** [comment]
I think documenting the recommended approach for webhook signature verification would be really helpful, even if `rawBody` isn't added to core.

A simple example in the Guides section showing how to use a `preParsing` hook to capture the raw body for specific routes would cover the most common use case:

```js
fastify.post('/webhook', {
  preParsing: async (request, reply, payload) => {
    const chunks = []
    for await (const chunk of payload) {
      chunks.push(chunk)
    }
    const rawBody = Buffer.concat(chunks)
    request.rawBody = rawBody
    
    // Return a new readable stream for the JSON parser
    const { Readable } = require('node:stream')
    return Readable.from(rawBody)
  }
}, async (request) => {
  // Verify webhook signature using request.rawBody
  const isValid = verifySignature(request.rawBody, request.headers['x-signature'])
  // request.body is still parsed as JSON
})
```

This is a pattern I've had to figure out myself and I imagine many Stripe/GitHub webhook integrators run into the same issue. Would a docs PR for this be welcome?

**Comment by Ilia01** [comment]
I think the missing docs piece here is not just "how do I get rawBody?", it is "how do I do signature verification on one webhook route without changing parser behavior everywhere else?"

For Stripe, GitHub, Svix, etc., the verifier has to see the exact bytes that arrived on the socket before JSON parsing or re-serialization changes anything. Once the body has been parsed first, `JSON.stringify(...)` is not a safe reconstruction of the signed payload.

So the most useful docs/example here would be a route- or plugin-scoped pattern like:

```ts
fastify.register(async function webhookRoutes(app) {
  app.addContentTypeParser(
    'application/json',
    { parseAs: 'buffer' },
    (_req, body, done) => done(null, body),
  )

  app.post('/stripe/webhook', async (request, reply) => {
    const rawBody = request.body as Buffer
    const signature = request.headers['stripe-signature'] as string

    const event = stripe.webhooks.constructEvent(
      rawBody,
      signature,
      process.env.STRIPE_WEBHOOK_SECRET!,
    )

    reply.code(200).send({ received: true })
  })
})
```

That keeps the raw-body cost scoped to the webhook routes that actually need it.

I ended up building HookLens because this exact problem kept surfacing as "wrong secret" when the real bug was "the framework changed the body before verification." Even if this stays out of core and in docs/plugin territory, I think Fastify would benefit from a first-party webhook-signature example that makes this route-scoped pattern explicit.

**Comment by Ilia01** [comment]
Strongly agree the docs should call this out explicitly.

The recurring trap with webhook verification is that the request looks correct in logs, the secret looks correct, and verification still fails because the framework already handed you a parsed/reconstructed body. For signature checks, an equivalent JSON object is not enough. The relevant input is the exact bytes that arrived on the wire.

So route-level `rawBody` guidance would save people a lot of dead-end secret rotation.

I kept running into exactly this pattern with Stripe and GitHub, which is why I built HookLens: capture before parsing, see whether the verifier failed because of missing header / timestamp / signature mismatch / likely body mutation, then replay the same delivery locally after fixing middleware.

Even without that tool, I think the docs should make the order explicit:
1. preserve raw body first
2. verify signature second
3. parse after that
